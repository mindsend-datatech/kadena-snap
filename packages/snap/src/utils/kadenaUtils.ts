import { SLIP10Node } from '@metamask/key-tree';
import type { DerivedAccount } from '../types';
import { BASE_ACCOUNT_NAME } from './constants';

/**
 * Get the key deriver for Kadena using BIP-32 entropy.
 * @returns The SLIP10Node for Kadena key derivation.
 */
export async function getKeyDeriver(): Promise<SLIP10Node> {
  const kadenaNode = await snap.request({
    method: 'snap_getBip32Entropy',
    params: {
      path: ['m', "44'", "626'"],
      curve: 'ed25519',
    },
  });

  const kadenaSlip10Node = await SLIP10Node.fromJSON(kadenaNode);
  return kadenaSlip10Node;
}

export type GetKeysFromIndexParams = {
  index: number;
};

/**
 * Derive account information from a given index.
 * @param keyDeriver - The SLIP10Node for key derivation.
 * @param index - The account index to derive.
 * @returns The derived account information.
 */
export const getAccountsFromIndex = async (
  keyDeriver: SLIP10Node,
  index: number,
): Promise<DerivedAccount> => {
  const derived = await keyDeriver.derive([`slip10:${index}'`]);
  return {
    index,
    address: derived.publicKey.replace(/^0x00/, 'k:'),
    name: `${BASE_ACCOUNT_NAME} ${index + 1}`,
    privateKey: derived.privateKey,
    publicKey: derived.publicKey,
  };
};
