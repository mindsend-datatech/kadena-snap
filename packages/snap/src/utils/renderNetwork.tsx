import type { Network } from '../types/index';
import { Text, Bold, Box } from '@metamask/snaps-sdk/jsx';

export default function renderNetwork(network: Network) {
  const {
    name,
    networkId,
    nodeUrl,
    isTestnet,
    transactionListTtl,
    transactionListUrl,
    blockExplorerAddress,
    blockExplorerTransaction,
    blockExplorerAddressTransactions,
  } = network;
  
  return (
    <Box>
      <Text>
        <Bold>Network name: {name}</Bold>
      </Text>
      <Text>Network ID: {networkId}</Text>
      <Text>Testnet? {String(isTestnet ? "yes" : "no")}</Text>
      <Text>Node URL: {nodeUrl}</Text>
      <Text>Transaction list TTL: {String(transactionListTtl)}</Text>
      <Text>Transaction list URL: {transactionListUrl}</Text>
      <Text>Explorer Address URL: {blockExplorerAddress}</Text>
      <Text>Explorer Transaction URL: {blockExplorerTransaction}</Text>
      <Text>Explorer Address Transactions URL: {blockExplorerAddressTransactions}</Text>
    </Box>
  );
}