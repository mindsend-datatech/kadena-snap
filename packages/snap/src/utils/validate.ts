import { InternalError, InvalidParamsError } from '@metamask/snaps-sdk';

type DefinitionValue = "string" | "number" | "boolean" | "object";

export type ValidateDefinition = {
  [key: string]: DefinitionValue;
};

type Obj = Record<string, unknown>

/**
 * Create a validator function for a given validation definition.
 * @param def - The validation definition specifying expected types.
 * @returns A function that validates objects against the definition.
 */
export function makeValidator<ObjectType extends Obj>(def: ValidateDefinition): (obj: ObjectType) => void {
  return (obj: ObjectType) => validateParams(def, obj);
}

/**
 * Validate an object against a validation definition.
 * @param def - The validation definition specifying expected types.
 * @param obj - The object to validate.
 * @throws InvalidParamsError if validation fails.
 */
export function validateParams<ObjectType extends Obj>(def: ValidateDefinition, obj: ObjectType): void {
  const errors = [];
  for(const [key, expectedType] of Object.entries(def)) {
    if (!(validateField(obj[key], expectedType))) {
      errors.push(key);
    }
  }

  if (errors.length) {
    const fieldMessage = errors.map(key => {
      const expectedType = def[key];
      const suffix = expectedType.endsWith('[]') ? '' : ` received ${typeof obj[key]}`;
      return `"${key}: expected ${expectedType}${suffix}"`
    }).join(', ');

    const prefix = `${errors.length} parameter validation error${errors.length > 1 ? 's':''}: `;

    throw new InvalidParamsError(`${prefix}${fieldMessage}`);
  }
}

/**
 * Validate a single field against its expected type.
 * @param value - The value to validate.
 * @param defValue - The expected type definition.
 * @returns True if validation passes, false otherwise.
 */
function validateField(value: unknown, defValue: DefinitionValue): boolean {
  switch(defValue) {
    case 'boolean':
      return validateBoolean(value);
    case 'number':
      return validateNumber(value);
    case 'string':
      return validateString(value);
    case 'object':
      return validateObject(value);
    default:
      throw new InternalError(`Unexpected validation type: ${defValue as string}`);
  }
}

/**
 * Validate that a value is a boolean.
 * @param n - The value to validate.
 * @returns True if the value is a boolean.
 */
function validateBoolean(n: unknown): n is boolean {
  return typeof n === "boolean";
}

/**
 * Validate that a value is a finite number.
 * @param n - The value to validate.
 * @returns True if the value is a finite number.
 */
function validateNumber(n: unknown): n is number {
  return typeof n === "number" && Number.isFinite(n);
}

/**
 * Validate that a value is a string.
 * @param s - The value to validate.
 * @returns True if the value is a string.
 */
function validateString(s: unknown): s is string {
  return typeof s === "string";
}

/**
 * Validate that a value is an object.
 * @param arr - The value to validate.
 * @returns True if the value is an object.
 */
function validateObject(arr: unknown): arr is object {
  return typeof arr === "object" && arr !== null;
}
