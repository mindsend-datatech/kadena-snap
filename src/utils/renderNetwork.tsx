import type { Network } from '../types/index';
import { Text, Bold } from '@metamask/snaps-sdk/jsx';

export default function renderNetwork(network: Network) {
  const {
    name,
    networkId,
    nodeUrl,
    isTestnet,
    transactionListTtl,
    transactionListUrl,
    blockExplorerAddress,
    blockExplorerTransaction,
    blockExplorerAddressTransactions,
  } = network;
  
  return [
    <Text key="name">
      <Bold>Network name: {name}</Bold>
    </Text>,
    <Text key="networkId">Network ID: {networkId}</Text>,
    <Text key="isTestnet">Testnet? {String(isTestnet ? "yes" : "no")}</Text>,
    <Text key="nodeUrl">Node URL: {nodeUrl}</Text>,
    <Text key="transactionListTtl">Transaction list TTL: {String(transactionListTtl)}</Text>,
    <Text key="transactionListUrl">Transaction list URL: {transactionListUrl}</Text>,
    <Text key="blockExplorerAddress">Explorer Address URL: {blockExplorerAddress}</Text>,
    <Text key="blockExplorerTransaction">Explorer Transaction URL: {blockExplorerTransaction}</Text>,
    <Text key="blockExplorerAddressTransactions">Explorer Address Transactions URL: {blockExplorerAddressTransactions}</Text>,
  ];
}