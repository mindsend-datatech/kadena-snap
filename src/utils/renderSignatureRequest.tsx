import { parseCapabilities } from "../utils/parseCapabilities";
import { Divider, Text, Bold, Box } from "@metamask/snaps-sdk/jsx";
import type { ISigner, ICommandPayload } from "@kadena/types";
import type { ApiParams } from "../types";

const TXN_LIFETIME_WARN_SEC = 3600;

function renderLifetime(txn: ICommandPayload) {
  const creationDate = new Date(txn.meta.creationTime * 1000);
  const expiryDate = new Date(creationDate.getTime() + txn.meta.ttl * 1000);

  // expiry interval in ms - counting from "now" (not creationDate)
  const expiryIntervalMs = expiryDate.getTime() - new Date().getTime();

  // minutes or hours until expiry, ceiling'ed, for display
  const expiryHours = Math.ceil(expiryIntervalMs / (1000 * 3600));
  const expiryMinutes =
    expiryHours === 1 ? Math.ceil(expiryIntervalMs / (1000 * 60)) : null;
  let labelTTL =
    expiryMinutes === null
      ? `${expiryHours} hour${expiryHours === 1 ? "" : "s"}`
      : `${expiryMinutes} minute${expiryMinutes === 1 ? "" : "s"}`;

  // warn if transaction expires in more than 1 hour
  const ttlWarnings = [];
  if (expiryIntervalMs > 1000 * TXN_LIFETIME_WARN_SEC) {
    ttlWarnings.push(`⚠️  Unusually long transaction lifetime`);
  }

  if (expiryIntervalMs < 0) {
    ttlWarnings.push(`⚠️  Transaction already expired`);
    labelTTL = `**Expired**`;
  }

  const lifetimeElements = [
    <Text key="lifetime-label">Transaction lifetime:</Text>,
    <Text key="lifetime-value">
      {expiryIntervalMs < 0 ? <Bold>Expired</Bold> : labelTTL} (expires{" "}
      {expiryDate.toLocaleString()})
    </Text>,
    ...ttlWarnings.map((warning, index) => (
      <Text key={`warning-${index}`}>{warning}</Text>
    )),
  ];

  return lifetimeElements;
}

export function renderTransactionRequest(
  signer: ISigner,
  txn: ICommandPayload,
  snapApi: ApiParams,
) {
  const accounts = Object.values(snapApi.state.accounts).reduce(
    (memo, account) => {
      memo[account.address] = account.name;
      return memo;
    },
    {} as Record<string, string>,
  );
  const capStrings = parseCapabilities(signer, txn, accounts);

  const capElements = capStrings.flatMap((str, idx) => {
    const total = capStrings.length;
    const lines = str.split("\n");
    return [
      <Text key={`cap-${idx}-title`}>
        <Bold>
          APPROVING ({String(idx + 1)}/{String(total)})
        </Bold>
      </Text>,
      ...lines.map((line, lineIdx) => (
        <Text key={`cap-${idx}-line-${lineIdx}`}>{line}</Text>
      )),
      <Divider key={`cap-${idx}-divider`} />,
    ];
  });

  return (
    <Box>
      {capElements}
      {renderLifetime(txn)}
    </Box>
  );
}
