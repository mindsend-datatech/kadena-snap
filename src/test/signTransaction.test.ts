import { installSnap } from "@metamask/snaps-jest";
import { assert } from '@metamask/utils';
import { getAccounts } from "./helpers/test-utils";

describe("kda_signTransaction", () => {
  it("should get the signature of a transaction", async () => {
    const { request } = await installSnap();

    const accounts = await getAccounts(request);

    const response = request({
      method: "kda_signTransaction",
      params: {
        id: accounts[0].id,
        transaction: `{"payload":{"exec":{"code":"(coin.transfer-create \\"k:62bb7cf156ccfbe17bd6ca5460098ca9398a4aa3f04bd617f7a721b6e2e5aac7\\" \\"k:10a3f4a9e5317c8dba58435dfdd4121bd3e4b0483993e67b65bea9c3c1113af4\\" (read-keyset \\"ks\\") 0.100000000000)","data":{"ks":{"keys":["10a3f4a9e5317c8dba58435dfdd4121bd3e4b0483993e67b65bea9c3c1113af4"],"pred":"keys-all"}}}},"nonce":"kjs:nonce:1699369566001","signers":[{"pubKey":"62bb7cf156ccfbe17bd6ca5460098ca9398a4aa3f04bd617f7a721b6e2e5aac7","scheme":"ED25519","clist":[{"name":"coin.TRANSFER","args":["k:62bb7cf156ccfbe17bd6ca5460098ca9398a4aa3f04bd617f7a721b6e2e5aac7","k:10a3f4a9e5317c8dba58435dfdd4121bd3e4b0483993e67b65bea9c3c1113af4",{"decimal":"0.1"}]},{"name":"coin.GAS","args":[]}]}],"meta":{"gasLimit":2500,"gasPrice":1e-8,"sender":"k:62bb7cf156ccfbe17bd6ca5460098ca9398a4aa3f04bd617f7a721b6e2e5aac7","ttl":28800,"creationTime":1699369566,"chainId":"1"},"networkId":"mainnet01"}`,
      },
    });

    const ui = await response.getInterface({ timeout: 50000 });
    expect(JSON.parse(JSON.stringify(ui.content.props))).toMatchSnapshot();
    assert(ui.type === 'confirmation');
    await ui.ok();

    const result = await response;

    if (
      "result" in result.response &&
      typeof result.response.result === "string"
    ) {
      const parsedResult = JSON.parse(result.response.result);

      expect(parsedResult).toMatchObject({
        responses: [
          {
            commandSigData: {
              cmd: expect.stringContaining('"payload":{"exec":'),
              sigs: [
                {
                  pubKey:
                    "62bb7cf156ccfbe17bd6ca5460098ca9398a4aa3f04bd617f7a721b6e2e5aac7",
                  sig: "d51f64d0837faff9349ae877a19ea4d8d85f2c979d67294f32564bcbdbf79b449e80d3e132e21456c399096b1613f794b23e9e87d50238e78e457cbb755ef503",
                },
              ],
            },
            outcome: {
              hash: "_aK1VdurS1aeWEZPtkbHoaCjnBHD5Lqt4V0q-_FKYf4",
              result: "success",
            },
          },
        ],
      });
    } else {
      throw new Error(
        `Expected result string but got: ${JSON.stringify(result.response)}`
      );
    }
  });

  it("should throw an expection when payload is wrong", async () => {
    const { request } = await installSnap();

    const accounts: any = await request({
      method: "kda_getAccounts",
    });

    const response = await request({
      method: "kda_signTransaction",
      params: {
        id: accounts.response.result[0].id,
        transaction: `{"meta":{"gasLimit":2500,"gasPrice":1e-8,"sender":"k:599b9ccec365e813031b49a558d86f66e9bb210084cea4f59f57c91bd454f0f8","ttl":28800,"creationTime":1699369566,"chainId":"1"},"networkId":"mainnet01"}`,
      },
    });

    const responseString = JSON.stringify(response);
    expect(responseString.includes("error")).toBe(true);
    expect(responseString.includes(`"code":-32603`)).toBe(true);
  });
});
