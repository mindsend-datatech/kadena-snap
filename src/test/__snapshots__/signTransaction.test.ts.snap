// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kda_signTransaction should get the signature of a transaction 1`] = `
{
  "children": [
    {
      "key": null,
      "props": {
        "children": "Transaction signature request",
      },
      "type": "Text",
    },
    {
      "key": null,
      "props": {},
      "type": "Divider",
    },
    [
      {
        "key": "cap-0-title",
        "props": {
          "children": {
            "key": null,
            "props": {
              "children": [
                "APPROVING (",
                "1",
                "/",
                "2",
                ")",
              ],
            },
            "type": "Bold",
          },
        },
        "type": "Text",
      },
      {
        "key": "cap-0-line-0",
        "props": {
          "children": "Send: 0.1 KDA",
        },
        "type": "Text",
      },
      {
        "key": "cap-0-line-1",
        "props": {
          "children": "From:",
        },
        "type": "Text",
      },
      {
        "key": "cap-0-line-2",
        "props": {
          "children": "k:62bb7cf156ccfbe17bd6ca5460098ca9398a4aa3f04bd617f7a721b6e2e5aac7 (Kadena Account 1) (chain 1)",
        },
        "type": "Text",
      },
      {
        "key": "cap-0-line-3",
        "props": {
          "children": "To:",
        },
        "type": "Text",
      },
      {
        "key": "cap-0-line-4",
        "props": {
          "children": "k:10a3f4a9e5317c8dba58435dfdd4121bd3e4b0483993e67b65bea9c3c1113af4 (chain 1)",
        },
        "type": "Text",
      },
      {
        "key": "cap-0-divider",
        "props": {},
        "type": "Divider",
      },
      {
        "key": "cap-1-title",
        "props": {
          "children": {
            "key": null,
            "props": {
              "children": [
                "APPROVING (",
                "2",
                "/",
                "2",
                ")",
              ],
            },
            "type": "Bold",
          },
        },
        "type": "Text",
      },
      {
        "key": "cap-1-line-0",
        "props": {
          "children": "Gas spend:",
        },
        "type": "Text",
      },
      {
        "key": "cap-1-line-1",
        "props": {
          "children": "Up to 0.000025 KDA",
        },
        "type": "Text",
      },
      {
        "key": "cap-1-divider",
        "props": {},
        "type": "Divider",
      },
      {
        "key": "lifetime-label",
        "props": {
          "children": "Transaction lifetime:",
        },
        "type": "Text",
      },
      {
        "key": "lifetime-value",
        "props": {
          "children": [
            {
              "key": null,
              "props": {
                "children": "Expired",
              },
              "type": "Bold",
            },
            " (expires ",
            "07/11/2023, 20:06:06",
            ")",
          ],
        },
        "type": "Text",
      },
      {
        "key": "warning-0",
        "props": {
          "children": "⚠️  Transaction already expired",
        },
        "type": "Text",
      },
    ],
  ],
}
`;
